### ISSUE TITLE
Wrong video width

### ISSUE THREAD

**Issue by Sep95** [ZeBobo5/Vlc.DotNet#508]
I have an issue  about Vlc.DotNet.

# Generic information

- **Vlc.DotNet version** : 3.0.0.0
- **Vlc.DotNet project used** : WPF
- **libvlc version** : 3.0.3.0 x86 via official VideoLAN homepage
- **.net version** : 4.5
- **Project language** : C#
- **Project build architecture** : AnyCPU (prefer 32bits)
- **Operating system** : Windows 10 (x64)

# Summary

I'm using the new `ImageSource` provider for binding the movie to an `Image` control. When rendering a Full HD movie in an Full HD sized `Image` control (Width:1920px, Height: 1080px), I have approx. 10px border on the left and right hand side. So correct height, but incorrect width. I don't know why.

So I used the UI-Debugger for XAML to determine the actual sizing while runtime. The property `RenderSize` had the value: 1902.38532110092,1080. So correct height, but incorrect width. If I decrease the width of the `Image` control around 20px, the border disappears and everything works fine.

Is there maybe an issue with calculating the output size for the `ImageSource` property? In my layout, the `Image` fills the whole window with in every scenario (verified with UI-Debugger for XAML), but the size of video output in the given case not. I use 100% scaling on my screen, so that's no reason.

I confirm that my issue doesn't happen in VLC itself.

**Comment by jeremyVignelles** [comment]
What size does VLC show in the media properties?

It's weird that the control doesn't have the correct size. If you don't want the border, you can create an Image control that you resize yourself, as described here : https://github.com/ZeBobo5/Vlc.DotNet/tree/develop/src/Samples/Samples.Wpf.Image

Please also note that if perf is critical for you, I'd recommend using the WinForms control instead.

**Comment by Sep95** [comment]
VLC shows the correct size (1920 x 1080). I also tried other Full HD videos, but without success. 

I can compile the Samples.WPF.Image project, but it unfortunately crashes on startup with an  
`System.ComponentModel.Win32Exception` ("%1 ist keine zulässige Win32-Anwendung"). I've already adapted the libvlc path to the newest x86 version via nuget.

**Comment by jeremyVignelles** [comment]
Samples.WPF.Image is built for x86 if you give it x86 path?

**Comment by Sep95** [comment]
I've just forced the sample to use x86 path. But still the same error.

**Comment by jeremyVignelles** [comment]
As far as I remember, I configured this sample as `AnyCPU` (not `AnyCPU (prefer 32 bits`) this means that the process will be run in x64 mode if the OS supports this, and that you must use the x64 libvlc.

If you want to use the x86 libvlc, you will need to change that in the project build properties to x86 or AnyCPU (prefer 32 bits)

**Comment by Sep95** [comment]
Thanks! Now the sample works, but there is the same issue. Here are steps for reproducing the issue with the `Wpf.Image` sample:
1. Change the video source to an Full HD video. E.g.
`this.sourceProvider.MediaPlayer.Play(new Uri("http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4"));`
2. Fix the size of video Image to Full HD for easier reproducing:
`<Image x:Name="Video" Stretch="Uniform" Width="1920" Height="1080"/>`
3. Run the program. If the windows gets resized to a width of 1920px, you can see the following border (part of the blurry blackground):
![grafik](https://user-images.githubusercontent.com/32958359/51073911-c4c9f580-1677-11e9-917e-013d785a6ca4.png)

**Comment by jeremyVignelles** [comment]
I tried your example and saw that libvlc tells the `VlcVideoSourceProvider` that the video is 1920x1088 instead of 1080.
I don't think I can really do something about it. You can try:
- to play another video (maybe with another video codec) and see if you're having the same issue.
- to reimplement the VlcVideoSourceProvider and set explicitely the height to 1080.
- to use the WinForms control instead, you'd get better perfs.
- to ask the question to videolan directly.

I'm sorry I can't help you more about that.

**Comment by Sep95** [comment]
I found the issue. VLC distinguishes _buffer size_ from _display size_ (see [here](https://forum.videolan.org/viewtopic.php?t=127336)). In case of the _Big buck bunny_ 1080p video, the buffer size is 1920 x 1088 and the display size is 1920 x 1080:

![grafik](https://user-images.githubusercontent.com/32958359/52080398-f0b20a00-2597-11e9-83c2-57c258c290ed.png)

Can the `VlcVideoSourceProvider` be changed, to get the display size from VLC instead of the buffer size?

**Comment by jeremyVignelles** [comment]
No, it can't , see https://github.com/ZeBobo5/Vlc.DotNet/blob/fb42e9a59950fcc27738d0f36477812727ee2152/src/Vlc.DotNet.Wpf/VlcVideoSourceProvider.cs#L145
This method is called by libvlc.

The only thing we can do is to open an issue on vlc's side. I'll reopen this to track the task of submitting a bugreport on vlc's trac

**Comment by Sep95** [comment]
I've added an [entry](https://forum.videolan.org/viewtopic.php?f=32&t=148144) to the official VLC forum. But no response so far.

**Comment by Sep95** [comment]
VLC seems to provide information on the video track (like the correct size) before the VideoFormat(...) callback is entered (see [here](https://forum.videolan.org/viewtopic.php?f=32&t=148144)). Unfortunately I don't find them.

**Comment by jeremyVignelles** [comment]
If the only way is to disregard the parameter, we should replace these two lines:

https://github.com/ZeBobo5/Vlc.DotNet/blob/fb42e9a59950fcc27738d0f36477812727ee2152/src/Vlc.DotNet.Wpf/VlcVideoSourceProvider.cs#L150-L151

With something that uses the width and height of the media

**Comment by Sep95** [comment]
Thats right. But I don't know where to find the correct size (using VLC) before _VideoFormat()_ callback is entered.

**Comment by jeremyVignelles** [comment]
You will need to enumerate the tracks, find the first video track and get width and height there. see :
https://github.com/ZeBobo5/Vlc.DotNet/blob/f0d3e2aca37500ccd2710c8b9b286ccc485f1636/src/Samples/Samples.WinForms.Advanced/Sample.cs#L146

**Comment by Sep95** [comment]
That works fine! I'm using the following code in WPF enviroment (called in the `LengthChanged` event, that means after the video is loaded) for determining the correct size of the first Videotrack:
```
var md = vlcSrcProvider.MediaPlayer.GetMedia();
var track = md.Tracks[0].TrackInfo as Vlc.DotNet.Core.Interops.VideoTrack;
uint trackWidth = track.Width;
uint trackHeight = track.Height;
```
But we get this information to late to correct `pitches` and `lines` (far after the `VideoFormat()` is called).

**Comment by jeremyVignelles** [comment]
Did you try to integrate that in the VideoFormat callback?

**Comment by Sep95** [comment]
I've just tried it out. It works if I add the following code before calculating _pitches_ and _lines_ in the `VideoFormat()` callback:

```
var md = MediaPlayer.GetMedia();
for (int i = 0; i < md.Tracks.Length; i++)
{
	if (md.Tracks[i].Type == MediaTrackTypes.Video)
	{
		var track = md.Tracks[i].TrackInfo as VideoTrack;
		width = track.Width;
		height = track.Height;
		break;
	}
}
```

The only problem: I cannot determine, which of the tracks is currently played. So if you have multiple tracks in a playlist, my code would only get the correct size for the first videotrack.

**Comment by jeremyVignelles** [comment]
That's not a problem for me : If I remember correctly, you can in theory have multiple video tracks per file, but in practice, you can only display the first video track with libvlc.

I don't know if there is any way to play another track or if that means anything at all.

Can you make a PR for that? your code seems good but could be simplified by using a foreach

**Comment by Sep95** [comment]
Unfortunately I cannot create a branch for submitting the pull request. Is my account maybe blocked for doing so?

**Comment by jeremyVignelles** [comment]
You need to fork the project, make your branch there and make a PR

**Comment by Sep95** [comment]
Ok, thanks! I've just created the [pull request](https://github.com/ZeBobo5/Vlc.DotNet/pull/526).

**Comment by jeremyVignelles** [comment]
Closed by #526
